File: algorithms/234. Palindrome Linked List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast, reverse = head, None
        while fast and fast.next:
            fast = fast.next.next
            reverse, reverse.next, head = head, reverse, head.next
        fast = head.next if fast else head
        p = True
        while reverse:
            p = p and reverse.val == fast.val
            fast = fast.next
            head, head.next, reverse = reverse, head, reverse.next
        return p

File: algorithms/234. Palindrome Linked List/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        // @StefanPochmann
        ListNode * reverse = NULL, * fast = head, *tmp;
        while ( fast && fast->next ) {
            fast = fast->next->next;
            tmp = head->next;
            head->next = reverse;
            reverse = head;
            head = tmp;
        }
        fast = fast ? head->next : head;
        bool p = true;
        while ( reverse ) {
            p = p && reverse->val == fast->val;
            fast = fast->next;
            tmp = reverse->next;
            reverse->next = head;
            head = reverse;
            reverse = tmp;
        }
        return p;
    }
};
